### ISSUE TITLE
Heatmap error with MySQL

### ISSUE THREAD

**Issue by bogdanp05** [flask-dashboard/Flask-MonitoringDashboard#103]
When trying to see the heatmap while using MySQL, I get the following error:
```
127.0.0.1 - - [03/May/2018 02:07:57] "GET /dashboard/measurements/heatmap HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/bogdan/ENV/lib/python3.5/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/bogdan/Flask-MonitoringDashboard/flask_monitoringdashboard/core/auth.py", line 37, in wrapper
    return func(*args, **kwargs)
  File "/home/bogdan/Flask-MonitoringDashboard/flask_monitoringdashboard/views/dashboard/heatmap.py", line 21, in heatmap
    return render_template('fmd_dashboard/graph.html', form=form, graph=get_heatmap(form), title=TITLE)
  File "/home/bogdan/Flask-MonitoringDashboard/flask_monitoringdashboard/views/dashboard/heatmap.py", line 43, in get_heatmap
    for d in get_num_requests(db_session, end, form.start_date.data, form.end_date.data):
  File "/home/bogdan/Flask-MonitoringDashboard/flask_monitoringdashboard/database/endpoint.py", line 25, in get_num_requests
    group_by('newTime').all()
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2737, in all
    return list(self)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2889, in __iter__
    return self._execute_and_instances(context)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2912, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 948, in execute
    return meth(self, multiparams, params)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
    exc_info
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
    raise value.with_traceback(tb)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 507, in do_execute
    cursor.execute(statement, parameters)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/cursors.py", line 165, in execute
    result = self._query(query)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/cursors.py", line 321, in _query
    conn.query(q)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/connections.py", line 860, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/connections.py", line 1061, in _read_query_result
    result.read()
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/connections.py", line 1349, in read
    first_packet = self.connection._read_packet()
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/connections.py", line 1018, in _read_packet
    packet.check_error()
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/connections.py", line 384, in check_error
    err.raise_mysql_exception(self._data)
  File "/home/bogdan/ENV/lib/python3.5/site-packages/pymysql/err.py", line 107, in raise_mysql_exception
    raise errorclass(errno, errval)
sqlalchemy.exc.InternalError: (pymysql.err.InternalError) (1305, 'FUNCTION flask_test.strftime does not exist') [SQL: 'SELECT strftime(%(strftime_1)s, `functionCalls`.time) AS `newTime`, count(`functionCalls`.execution_time) AS count \nFROM `functionCalls` \nWHERE `functionCalls`.time >= %(time_1)s AND `functionCalls`.time <= %(time_2)s GROUP BY `newTime`'] [parameters: {'strftime_1': '%Y-%m-%d %H:00:00', 'time_2': datetime.datetime(2018, 5, 3, 23, 59, 59), 'time_1': datetime.datetime(2018, 4, 13, 0, 0)}] (Background on this error at: http://sqlalche.me/e/2j85)
```
The problem doesn't appear with SQLite, so we're using some db-specific code, probably `strftime`.

**Comment by bogdanp05** [comment]
Hi, Harold! Thanks for the tip, will look into it.

**Comment by Harold2017** [comment]
Hi, i made a trial and it may help in the following link's notebook cell #41:
[](https://github.com/Harold2017/tests/blob/master/sqlalchemy.ipynb)
I'm a beginner and please forgive my mass code XD

**Comment by marinswk** [comment]
Hello i'm getting something similar for /dashboard/measurements/versions as well using MySql

Exception on /dashboard/measurements/versions [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python2.7/dist-packages/flask_monitoringdashboard/core/auth.py", line 37, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/flask_monitoringdashboard/views/dashboard/versions.py", line 18, in page_boxplot_per_version
    return render_template('fmd_dashboard/graph.html', graph=get_boxplot_per_version(), title=TITLE)
  File "/usr/local/lib/python2.7/dist-packages/flask_monitoringdashboard/views/dashboard/versions.py", line 27, in get_boxplot_per_version
    versions = get_versions(db_session)
  File "/usr/local/lib/python2.7/dist-packages/flask_monitoringdashboard/database/versions.py", line 26, in get_versions
    order_by(asc(FunctionCall.time)).all()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 2737, in all
    return list(self)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 2889, in __iter__
    return self._execute_and_instances(context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 2912, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 948, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/default.py", line 507, in do_execute
    cursor.execute(statement, parameters)
  File "/usr/local/lib/python2.7/dist-packages/MySQLdb/cursors.py", line 205, in execute
    self.errorhandler(self, exc, value)
  File "/usr/local/lib/python2.7/dist-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
InterfaceError: (_mysql_exceptions.InterfaceError) (-1, 'error totally whack') [SQL: u'SELECT DISTINCT `functionCalls`.version AS anon_1 \nFROM `functionCalls` \nWHERE true = 1 ORDER BY `functionCalls`.time ASC'] (Background on this error at: http://sqlalche.me/e/rvf5)

**Comment by bogdanp05** [comment]
@Harold2017 Your code does 1 initial query to get the days, and then for every day it does another query to the db, which is not at all efficient. But I appreciate your initiative and if you want to contribute in the future you can just fork the repo.

@marinswk Your are right. At the moment, all the functionalities of the project are designed and tested using SQLite, so there are a few mismatches when using another db. I opened #116 to address all of these db-related errors.

I fixed this issue in #115 .

**Comment by marinswk** [comment]
No problems, i'm just happy that signaling bugs will help you improve the application, is a really cool project and you are keeping it up in a cool way.

The fact of fixing mysql issues might be convenient for many people, for example i prefer to use a remote db instead of having the dashboard writing in a local sqlite file :)
